src/functional/ffront/dialect_parser.py

```diff
@@ -67,30 +66,27 @@ def apply(
         closure_refs: ClosureRefs,
         externals: Optional[dict[str, Any]] = None,
     ) -> DialectRootT:
+
         source, filename, starting_line = source_definition
         try:
+            raw_ast = ast.parse(textwrap.dedent(source)).body[0]
             definition_ast = cls._preprocess_definition_ast(
-                ast.parse(textwrap.dedent(source)).body[0]
+                RewriteLineNumbers.apply(raw_ast, starting_line)
             )
             output_ast = cls._postprocess_dialect_ast(
                 cls(
-                    source=source,
-                    filename=filename,
-                    starting_line=starting_line,
+                    source_definition=source_definition,
                     closure_refs=closure_refs,
                     externals_defs=externals or {},
                 ).visit(definition_ast)
             )
             if __debug__:
                 _assert_source_invariants(source_definition, closure_refs)
         except SyntaxError as err:
-            # TODO(tehrengruber): Instead of enhancing the exception here and
-            #  having source information in the instance we could preprocess
-            #  the ast and offset the line numbers there.
             if not err.filename:
                 err.filename = filename
-            if not isinstance(err, DialectSyntaxError):
-                err.lineno = (err.lineno or 1) + starting_line - 1
+            if not err.text:
+                err.text = source
```

Comment:
Another suggestion: I don't know if it is clear when or why `err.filename` and `err.text` are expected to be empty, but if it is indeed clear, I would add a comment here to explain it.